clowder-framework/clowder2
# 1086 file UI component flag current files in datasets

## Aim 
To create an option in the file component viewer to only see the files in the current dataset

## Methodology

The difficulty behind the task is because the datasetID depends on which dataset the extractor is called from and therefore can not be hardcoded into the `extractor_info` json. In order to bypass, this we use the property [`uiSchema`](https://rjsf-team.github.io/react-jsonschema-form/docs/api-reference/uiSchema/) (that I have previously misunderstood) and `"ui:options"`.

 uiSchema is actually used to format a given input and is dependent on the parameter name. `ui:options` allows us to pass in the datasetID as a field. Since the uiSchema uses parameter names I use a for loop to iterate through the form's parameters and add to the uiSchema

```js
	// The for loop is used to pass the datasetId to a widget if it is a clowderFile Widget
	if (
		selectedExtractor &&
		selectedExtractor["properties"] &&
		selectedExtractor["properties"]["parameters"] &&
		selectedExtractor["properties"]["parameters"]["schema"]
	) {
		const parameters = selectedExtractor["properties"]["parameters"]["schema"];
		for (const key in parameters) {
			// Check if field "format" present and equals "clowderFile"
			if (parameters[key].format && parameters[key].format === "clowderFile") {
				uiSchema[key] = {
					"ui:widget": "clowderFile",
					"ui:options": {
						datasetId: datasetId,
					},
				};
			}
		}
	}
```

The rest of the changes are pretty straightforward. It involves passing in the datasetID and a flag to only show the dataset files or the entire filesystem.

## How to test

Run an extractor, I used [wordcount](https://github.com/clowder-framework/clowder2-demos/tree/main/word-cloud-extractor) , locally as in to do the necessary modifications:



  1. Add another parameter in `extractor-info.json` file,  in the `schema` field of  `parameters`, if using wordcount, you will have to add it [here](https://github.com/clowder-framework/clowder2-demos/blob/0ba84d236d98290db5b9e75f6749fed52eee857d/word-cloud-extractor/extractor_info.json#L38) 
  ```js
  "DATASET_FILE_ID": {
          "type": "string",
          "title": "Dataset file ID",
          "description": "The Clowder file id of the dataset file to be used for inference",
          "format": "clowderFile",
          "showOnlyDatasetFiles": true
        }
  ```

3. When you open the extractor and open the fileselector, you should only see the current folders and files in the dataset
  ![image](https://github.com/clowder-framework/clowder2/assets/16596733/748ab92f-d801-41bd-896d-82141e90e667)

4. Re-run the extractor but tis time set the field "showOnlyDatasetFiles" to false or remove it
5. You should then see the complete clowder filesystem
![image](https://github.com/clowder-framework/clowder2/assets/16596733/dd6d8a38-5576-4a21-8d62-ad70c4efbfc3)
